File: ruby/triangle/triangle.rb
class Triangle
  def initialize(sides)
    @sides = sides.sort
  end

  def equilateral?
    @sides.uniq.length == 1 if valid_triangle?
  end

  def isosceles?
    (@sides.uniq.length == 2 || equilateral?) if valid_triangle?
  end

  def scalene?
    @sides.uniq.length == 3 if valid_triangle?
  end

  private

  def valid_triangle?
    @sides.all?(&:positive?) & (@sides[0] + @sides[1] >= @sides[2])
  end
end

File: JavaScript/triangle/triangle.js
export class Triangle {
    constructor(a, b, c) {
        this.a = a;
        this.b = b;
        this.c = c;
        this.sides = new Array(this.a, this.b, this.c).sort((a, b) => a - b);
    }

    getUniqueSides() {
        return this.sides.filter((side, index, sides) => sides.indexOf(side) === index).length;
    }

    kind() {
        if (!this.validTriangle()) {
            throw new Error();
        }
        
        switch (this.getUniqueSides()) {
            case 1:
                return 'equilateral';
            case 2:
                return 'isosceles';
            case 3:
                return 'scalene';    
        }
    }

    validTriangle() {
        return this.sides.every(side => side > 0) && this.sides[0] + this.sides[1] >= this.sides[2];
    }
}
